File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head
        out = []
        while curr:
            out.append(curr.val)
            curr = curr.next
        out.reverse()

        if not out:
            return None

        res = ListNode(val=out[0])
        curr = res

        for x in out[1:]:
            curr.next = ListNode(val=x)
            curr = curr.next

        return res

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head, n) {
        let i = 1;
        let curr = head;
        // count starts at the end of the list

        while (curr.next) {
            curr = curr.next;
            i++;
        }

        if (i == n) {
            head = head.next;
            return head; 
        }

        let prev = null;
        curr = head;

        while (curr) {
            prev = curr;
            curr = curr.next;
            i--;

            if (i == n) {
                prev.next = curr.next;
                break;
            }
        }

        return head;
    }
}
